src/Infrastructure/LeanCode.SendGrid/SendGridException.cs

```diff
@@ -0,0 +1,78 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net;
+using System.Text.Json;
+using System.Threading.Tasks;
+using SendGrid;
+
+namespace LeanCode.SendGrid
+{
+    public class SendGridException : Exception
+    {
+        private static readonly JsonSerializerOptions SerializerOptions = new JsonSerializerOptions
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        };
+
+        public HttpStatusCode ResponseStatusCode { get; }
+        public ImmutableArray<string> ErrorMessages { get; } = ImmutableArray<string>.Empty;
+
+        private SendGridException(
+            HttpStatusCode responseStatusCode,
+            IEnumerable<SendGridError>? errors)
+            : base($"SendGrid request failed with status code {responseStatusCode}.")
+        {
+            ResponseStatusCode = responseStatusCode;
+
+            if (errors is object)
+            {
+                ErrorMessages = errors
+                    .SelectMany(EnumerateMessages)
```

Comment:
Why not a simple `Select(e => e.Message).Where(not null)`?

Reply:
Because I like to overthink such simple problems :smile:

Reply:
Also it will cause the type to be inferred as `IEnumerable<string?>`, which `SelectMany` nicely avoids.

Reply:
`!` to the rescue!